### PR TITLE
Eliminate need for parens around labeled tuple types

### PR DESCRIPTION
**Pull request by ccasin**
This is the third of three PRs that fix up parsing for labeled tuples to require fewer parens.  Note that it is into the `labeled-tuples` branch.

This one is a little different than the other two, because there is a lot more ambiguity here.  For example `x:int*bool -> int` as a function with a single labeled argument (which is an unlabeled tuple).  So this PR must carefully avoid parsing it as a function with a single unlabled argument (which is a labeled tuple).  If you want that type, you must write `(x:int*bool) -> int`, but we don't want to require parens around labeled tuple types in non-function-argument positions.

This ambiguity is solved by splitting the parsing into two cases: one for labeled tuples that don't have a label on the first element (which can be parsed in just the same way that tuples have always been parsed) and labeled tuples that do have a label on the first element, which I had to put somewhere else in the grammar to deal with these ambiguities

There are various other potential ambiguities. There is a particularly nasty corner case around module with type constraints, which I describe in detail in a comment so I won't here.  I solved this with some kind of gross duplication in the grammar.  There is probably a cleaner solution!  But I think it's probably not worth the effort to find it, unless it's obvious to someone.

Now that all three of these PRs are ready, I've set up a testing build of the internal code base to root out any corner cases where I've accidentally changed the parsing.

Review: @riaqn

### CODE REVIEW

**Review by riaqn — APPROVED** [review]
I verified that `parser.mly` unfolds correctly. I can't ensure that reduce/shift conflict are resolved in the correct way, but given that the expected tests are all passed, I guess they are fine.

File: ocaml/typing/oprint.ml [hunk 1]
```diff
@@ -457,10 +456,18 @@ let rec print_out_type_0 mode ppf =
...
+   - It is an argument to a function ([~arg])
+   - Or, there is at least one mode to print.
+ *)
+and print_out_type_mode ~arg mode ppf ty =
   let is_local = is_local mode in
   let is_unique = is_unique mode in
   let is_once = is_once mode in
+  let parens =
+    is_initially_labeled_tuple ty
+    && (arg || is_local || is_unique || is_once)
```
[the remark this answers is not in the document]
Comment (ccasin) [reply ocaml/typing/oprint.ml:469]: That's right.  This is a limitation of the parser.  This parses:
```ocaml
int -> local_ (x:int * y:bool)
```
But without the parens, it does not.  This seems fine to me, but let me know if you disagree.

**Comment by ccasin** [comment]
@riaqn I pushed a small additional commit to fix a bug I found while running this on jane.  The bug is just that a couple locations were wrong in the new productions.  Can you take a quick glance at it and see if you agree?

File: ocaml/typing/oprint.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (riaqn) [reply ocaml/typing/oprint.ml:469]: Yes let's keep the parser and printer consistent. Changin the parser seems like a lot of trouble, so let's keep it this way.

**Comment by riaqn** [comment]
@ccasin yes it looks right to me (sorry I didn't spot this earlier)

**Pull request merged**